packages/core/core/src/BundleGraph.js

```diff
@@ -348,22 +348,40 @@ export default class BundleGraph {
     // Remove bundle node if it no longer has any entry assets
     let bundleNode = nullthrows(this._graph.getNode(bundle.id));
     if (this._graph.getNodesConnectedFrom(bundleNode).length === 0) {
-      let bundleGroupNodes = this._graph.getNodesConnectedTo(
-        bundleNode,
-        'bundle',
+      this.removeBundle(bundle);
+    }
+
+    this._bundleContentHashes.delete(bundle.id);
+  }
+
+  removeBundle(bundle: Bundle) {
+    // Remove bundle node if it no longer has any entry assets
+    let bundleNode = nullthrows(this._graph.getNode(bundle.id));
+
+    let bundleGroupNodes = this._graph.getNodesConnectedTo(
+      bundleNode,
+      'bundle',
+    );
+    this._graph.removeNode(bundleNode);
+
+    // Remove bundle group node if it no longer has any bundles
+    for (let bundleGroupNode of bundleGroupNodes) {
+      invariant(bundleGroupNode.type === 'bundle_group');
+      let isBundleGroupEntry = bundle.entryAssetIds.includes(
+        bundleGroupNode.value.entryAssetId,
       );
-      this._graph.removeNode(bundleNode);
 
-      // Remove bundle group node if it no longer has any bundles
-      for (let bundleGroupNode of bundleGroupNodes) {
-        invariant(bundleGroupNode.type === 'bundle_group');
-        if (this._graph.getNodesConnectedTo(bundleGroupNode).length === 0) {
-          this._graph.removeNode(bundleGroupNode);
-        } else {
-          let index = bundleGroupNode.value.bundleIds.indexOf(bundle.id);
-          invariant(index >= 0);
-          bundleGroupNode.value.bundleIds.splice(index, 1);
-        }
+      if (
+        // If the bundle group's entry asset belongs to this bundle, the group
+        // was created because of this bundle. Delete the group.
+        isBundleGroupEntry ||
```

Comment:
@devongovett is this safe?

Removing the bundle node should remove any of its children, and removing the group should remove its siblings from the group. Neither of these cases should leave us in a state where we're removing access to assets that are depended on by other bundles, right?

Reply:
Also, should we recursively remove child bundles that aren't reachable by any others? We currently only ever call this when deduplicating, in which case these bundles _would_ be reachable, but there's nothing stopping this from being called in other cases. cc @padmaia @devongovett